core/app/models/spree/payment_method.rb

```diff
@@ -17,7 +17,10 @@ class PaymentMethod < Spree::Base
     scope :active, -> { where(active: true) }
     scope :available_to_users, -> { where(available_to_users: true) }
     scope :available_to_admin, -> { where(available_to_admin: true) }
-    scope :available_to_store, -> (store) { (store.present? && store.payment_methods.empty?) ? self : store.payment_methods }
+    scope :available_to_store, ->(store) do
+      raise ArgumentError, "You must provide a store" if store.nil?
+      store.payment_methods.empty? ? all : where(id: store.payment_method_ids)
+    end
```

Comment:
Thanks for the feedback!

Unfortunately `where(store: store)` doesn't work, because the payment method <-> store association is a `has_many :through`.

The reason for the conditional is to keep the existing, or at least expected, behaviour where if the store doesn't have any payment methods, we return all of them instead of an empty collection.

I agree the ArgumentError is not a drastic improvement. I discussed this with @jhawthorn on Friday and would be okay just letting it error the usual way too.